File: lksh/Z_Manaker_Prefix_Func/G.cpp
#include "bits/stdc++.h"

using namespace std;

#ifdef SEREGA
#include "../debug.h"
auto freopen_input_result_ = freopen("../input.txt", "r", stdin);
// auto freopen_output_result_ = freopen("output.txt", "r", stdout);
#endif

void solve() {
    string s = "#";
    char c;
    int n = 0;
    while (cin >> c) {
        // debugging purposes
        if (c == '$') break;
        n++;
        s += c;
        s += '#';
    }
    int l = 0, r = 0;
    vector<int> z(s.size());
    for (int i = 0; i <= s.size(); ++i) {
        z[i] = min(max(0, r - i), z[l + r - i]);
        while (z[i] <= i && i + z[i] < s.size() && s[i - z[i]] == s[i + z[i]]) {
            z[i]++;
        }
        if (z[i] + i > r) {
            l = i - z[i];
            r = i + z[i];
        }
    }
    long long ans = 0;
    for (int i : z) {
        ans += i;
    }
    cout << (ans - n - 1) / 2 - n;
}

signed main() {
    cin.tie(nullptr)->sync_with_stdio(false);
    solve();
}

File: rosatom/otbor/G.py
n, x = map(int, input().split())
a = list(zip(list(map(int, input().split())), list(map(int, input().split()))))
b = list(zip(list(map(int, input().split())), list(map(int, input().split()))))
c = list(zip(list(map(int, input().split())), list(map(int, input().split()))))
answer = 100000000000000000
for i in range(n):
    for j in range(n):
        for k in range(n):
            if a[i][1] + b[j][1] + c[k][1] <= x:
                answer = min(
                    answer,
                    max(a[i][0], b[j][0], c[k][0]) - min(a[i][0], b[j][0], c[k][0]),
                )
if answer == 100000000000000000:
    print(-1)
else:
    print(answer)
